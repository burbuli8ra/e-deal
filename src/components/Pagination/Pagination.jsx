import React from 'react';
import Styled from './Paginations.style';

const Pagination = ({
  currentPage,
  onClick,
  pageItems,
  totalItems
}) => {
  const totalPages = Math.ceil(totalItems / pageItems);
  const pages = Array.from({ length: totalPages }).map(
    (item, index) => ++index
  );
  const isFirstPage = currentPage === 1;
  const isLastPage = currentPage === totalPages;

  return (
    <Styled.Pagination data-testid="pagination">
      <Styled.PageItem
        data-testid="page-prev"
        isDisabled={isFirstPage}
        onClick={() => { if (!isFirstPage) onClick(--currentPage) }}
      >
        {'< Prev'}
      </Styled.PageItem>
      {pages.map(page => (
        <Styled.PageItem
          data-testid="page-item"
          isActive={currentPage === page}
          key={page}
          onClick={() => onClick(page)}
        >
          {page}
        </Styled.PageItem>
      ))}
      <Styled.PageItem
        data-testid="page-next"
        isDisabled={isLastPage}
        onClick={() => { if (!isLastPage) onClick(++currentPage) }}
      >
        {'Next >'}
      </Styled.PageItem>
    </Styled.Pagination>
  );
};

export default Pagination;