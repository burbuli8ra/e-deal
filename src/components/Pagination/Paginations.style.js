import styled from '@emotion/styled';
import colors from 'theme/colors';

const Pagination = styled.ul`
  align-self: flex-end;
  display: flex;
  flex: 0 1 100%;
  justify-content: center;
  list-style-type: none;
  margin-bottom: 10px;
  padding-left: 0;
`;
Pagination.displayName = 'Pagination';

const PageItem = styled.li`
  background-color: ${({ isActive, isDisabled }) =>
  // eslint-disable-next-line no-nested-ternary
    isDisabled ? colors.grey : (
      isActive ? colors.pastelBlue : colors.lightPastelBlue
    )};
  border-radius: 4px;
  color: ${colors.white};
  cursor: ${({ isDisabled }) => isDisabled ? 'not-allowed' : 'pointer'};
  margin-left: 4px;
  margin-right: 4px;
  padding: 10px;
  
  &:first-of-type {
    margin-left: 0;
  }
  
  &:last-of-type {
    margin-right: 0;
  }
`;
PageItem.displayName = 'PageItem';

export default {
  Pagination,
  PageItem
};