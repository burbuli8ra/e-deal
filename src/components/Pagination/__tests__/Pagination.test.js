import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import Pagination from '../Pagination';

const setup = props => <Pagination {...props} />;

describe('Pagination component', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = {
      currentPage: 1,
      onClick: jest.fn(),
      pageItems: 10,
      totalItems: 16
    };
    wrapper = render(setup(props));
  });

  test('should match snapshot', () => {
    const { asFragment } = wrapper;
    expect(asFragment()).toMatchSnapshot();
  });

  test('should not call previous page', () => {
    const { getByTestId } = wrapper;

    fireEvent.click(getByTestId('page-prev'));
    expect(props.onClick).not.toHaveBeenCalled();
  });

  test('should call previous page', () => {
    const updatedProps = {
      ...props,
      currentPage: 2
    };
    const { getByTestId, rerender } = wrapper;

    rerender(setup(updatedProps));

    fireEvent.click(getByTestId('page-prev'));
    expect(props.onClick).toHaveBeenCalledTimes(1);
  });

  test('should not call next page', () => {
    const updatedProps = {
      ...props,
      currentPage: 2
    };
    const { getByTestId, rerender } = wrapper;

    rerender(setup(updatedProps));

    fireEvent.click(getByTestId('page-next'));
    expect(props.onClick).not.toHaveBeenCalled();
  });

  test('should call next page', () => {
    const { getByTestId } = wrapper;

    fireEvent.click(getByTestId('page-next'));
    expect(props.onClick).toHaveBeenCalledTimes(1);
  });

  test('should call page', () => {
    const { getAllByTestId } = wrapper;

    fireEvent.click(getAllByTestId('page-item')[0]);
    expect(props.onClick).toHaveBeenCalledTimes(1);
  })
});