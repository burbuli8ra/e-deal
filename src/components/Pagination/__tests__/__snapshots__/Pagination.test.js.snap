// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component should match snapshot 1`] = `
<DocumentFragment>
  <ul
    class="css-1ofi96v"
    data-testid="pagination"
  >
    <li
      class="css-1559gqj"
      data-testid="page-prev"
    >
      &lt; Prev
    </li>
    <li
      class="css-1voxcuf"
      data-testid="page-item"
    >
      1
    </li>
    <li
      class="css-ln6vel"
      data-testid="page-item"
    >
      2
    </li>
    <li
      class="css-ln6vel"
      data-testid="page-next"
    >
      Next &gt;
    </li>
  </ul>
</DocumentFragment>
`;
